### ISSUE TITLE
Android: confirmation dialog is... unreable and crashes

### ISSUE THREAD

**Issue by nerzhul** [minetest/minetest#7122]
##### Issue type

 - Bug report

 
##### Minetest version

```
0.5.0 git
```

##### OS / Hardware

Operating system: Android 6.0

##### Summary

Here is the problem:

![37256354-bb4c1d36-2559-11e8-80af-a8b7b50aeb03](https://user-images.githubusercontent.com/3686677/45852276-d7ba6200-bd36-11e8-92c0-6a78b886b5ca.png)

Also when clicking on the dialog, MT crashes

##### Steps to reproduce

Build last master, connect to a 0.5 server and see MT crash

**Comment by jastevenson303** [comment]
If you just change the ok button to read "OK" then it should be okay.

**Comment by nerzhul** [comment]
@jastevenson303 the problem is not only the button but also the text... all should be upscaled

**Comment by srifqi** [comment]
What is the screen resolution and GUI scaling?

**Comment by srifqi** [comment]
Using Android 6.0.1, using APK debug build for my PR. Trying to connect to new server, it crashes (SIGSEGV) before loading dialog.
```
03-15 20:49:34.681 19970 19983 D Minetest: 2018-03-15 20:49:34: INFO[Main]: SCRIPTAPI: Initialized client game modules
03-15 20:49:34.681 19970 19983 D Minetest: 2018-03-15 20:49:34: INFO[Main]: Connecting to server at 23.28.87.79:30000
03-15 20:49:34.681 19970 19983 V Minetest: 2018-03-15 20:49:34: VERBOSE[Main]: Loading and running script *builtin*:init.lua
03-15 20:49:34.701 19970 19983 D Minetest: 2018-03-15 20:49:34: INFO[Main]: Client::peerAdded(): peer->id=1
03-15 20:49:34.711 19970 19983 D Minetest: 2018-03-15 20:49:34: INFO[Main]: Client packetcounter (20):
03-15 20:49:35.091 19970 20082 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionSend]: con(43/2440)RE-SENDING timed-out RELIABLE to 23.28.87.79(t/o=0.1): from_peer_id=2440, channel=0, seqnum=65501
03-15 20:49:35.191 19970 20082 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionSend]: con(43/2440)RE-SENDING timed-out RELIABLE to 23.28.87.79(t/o=0.1): from_peer_id=2440, channel=0, seqnum=65501
03-15 20:49:35.291 19970 20082 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionSend]: con(43/2440)RE-SENDING timed-out RELIABLE to 23.28.87.79(t/o=0.1): from_peer_id=2440, channel=0, seqnum=65501
03-15 20:49:35.381 19970 20083 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionReceive]: con(43/2440)WARNING: ACKed packet not in outgoing queue
03-15 20:49:35.481 19970 20083 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionReceive]: con(43/2440)WARNING: ACKed packet not in outgoing queue
03-15 20:49:35.581 19970 20083 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionReceive]: con(43/2440)WARNING: ACKed packet not in outgoing queue
03-15 20:49:44.511 19970 19983 E Minetest: 2018-03-15 20:49:44: ERROR[Main]: Connection timed out.
03-15 20:49:44.561 19970 20082 D Minetest: 2018-03-15 20:49:44: INFO[ConnectionSend]: con(43/2440)RunTimeouts(): Peer 1 has timed out. (source=peer->timeout_counter)
03-15 20:49:44.561 19970 20082 V Minetest: 2018-03-15 20:49:44: VERBOSE[ConnectionSend]: con(43/2440)RunTimeouts(): Removing peer 1
03-15 20:49:44.561 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: Audio: Deinitializing...
03-15 20:49:44.591 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: Audio: Deinitialized.
03-15 20:49:44.601 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: ~TextureSource() 25/20
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: Irrlicht resources after cleanup:
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 	Remaining meshes   : 0
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 	Remaining textures : 20
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		0:#DefaultFont
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		1:/storage/emulated/0/Minetest/textures/base/pack/blank.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		2:/storage/emulated/0/Minetest/textures/base/pack/blank.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		3:/storage/emulated/0/Minetest/textures/base/pack/server_flags_creative.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		4:/storage/emulated/0/Minetest/textures/base/pack/server_flags_creative.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		5:/storage/emulated/0/Minetest/textures/base/pack/server_flags_damage.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		6:/storage/emulated/0/Minetest/textures/base/pack/server_flags_damage.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		7:/storage/emulated/0/Minetest/textures/base/pack/server_flags_favorite.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		8:/storage/emulated/0/Minetest/textures/base/pack/server_flags_favorite.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		9:/storage/emulated/0/Minetest/textures/base/pack/server_flags_pvp.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		10:/storage/emulated/0/Minetest/textures/base/pack/server_flags_pvp.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		11:/storage/emulated/0/Minetest/textures/base/pack/server_ping_1.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		12:/storage/emulated/0/Minetest/textures/base/pack/server_ping_1.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		13:/storage/emulated/0/Minetest/textures/base/pack/server_ping_2.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		14:/storage/emulated/0/Minetest/textures/base/pack/server_ping_2.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		15:/storage/emulated/0/Minetest/textures/base/pack/server_ping_3.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		16:/storage/emulated/0/Minetest/textures/base/pack/server_ping_3.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		17:/storage/emulated/0/Minetest/textures/base/pack/server_ping_4.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		18:/storage/emulated/0/Minetest/textures/base/pack/server_ping_4.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		19:TTFontGlyphPage_Arimo.Regular.35_0
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 	Remaining materials: 24 (note: irrlicht doesn't support removing renderers)
03-15 20:49:44.611 19970 19983 F libc    : Fatal signal 11 (SIGSEGV), code 1, fault addr 0xd2 in tid 19983 (MinetestNativeT)
03-15 20:49:44.701  2185  2185 F DEBUG   : pid: 19970, tid: 19983, name: MinetestNativeT  >>> net.minetest.minetest <<<
03-15 20:49:45.581  3893  3893 E audit   : type=1701 msg=audit(1521121785.581:12860): auid=4294967295 uid=10179 gid=10179 ses=4294967295 subj=u:r:untrusted_app:s0:c512,c768 pid=19983 comm="MinetestNativeT" reason="memory violation" sig=11
```

**Comment by srifqi** [comment]
Using Android 6.0.1 and APK debug build for my PR. Trying to connect to new server, it crashes (SIGSEGV) **before** showing confirmation dialog.

Logcat:
```
03-15 20:49:34.681 19970 19983 D Minetest: 2018-03-15 20:49:34: INFO[Main]: SCRIPTAPI: Initialized client game modules
03-15 20:49:34.681 19970 19983 D Minetest: 2018-03-15 20:49:34: INFO[Main]: Connecting to server at 23.28.87.79:30000
03-15 20:49:34.681 19970 19983 V Minetest: 2018-03-15 20:49:34: VERBOSE[Main]: Loading and running script *builtin*:init.lua
03-15 20:49:34.701 19970 19983 D Minetest: 2018-03-15 20:49:34: INFO[Main]: Client::peerAdded(): peer->id=1
03-15 20:49:34.711 19970 19983 D Minetest: 2018-03-15 20:49:34: INFO[Main]: Client packetcounter (20):
03-15 20:49:35.091 19970 20082 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionSend]: con(43/2440)RE-SENDING timed-out RELIABLE to 23.28.87.79(t/o=0.1): from_peer_id=2440, channel=0, seqnum=65501
03-15 20:49:35.191 19970 20082 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionSend]: con(43/2440)RE-SENDING timed-out RELIABLE to 23.28.87.79(t/o=0.1): from_peer_id=2440, channel=0, seqnum=65501
03-15 20:49:35.291 19970 20082 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionSend]: con(43/2440)RE-SENDING timed-out RELIABLE to 23.28.87.79(t/o=0.1): from_peer_id=2440, channel=0, seqnum=65501
03-15 20:49:35.381 19970 20083 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionReceive]: con(43/2440)WARNING: ACKed packet not in outgoing queue
03-15 20:49:35.481 19970 20083 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionReceive]: con(43/2440)WARNING: ACKed packet not in outgoing queue
03-15 20:49:35.581 19970 20083 V Minetest: 2018-03-15 20:49:35: VERBOSE[ConnectionReceive]: con(43/2440)WARNING: ACKed packet not in outgoing queue
03-15 20:49:44.511 19970 19983 E Minetest: 2018-03-15 20:49:44: ERROR[Main]: Connection timed out.
03-15 20:49:44.561 19970 20082 D Minetest: 2018-03-15 20:49:44: INFO[ConnectionSend]: con(43/2440)RunTimeouts(): Peer 1 has timed out. (source=peer->timeout_counter)
03-15 20:49:44.561 19970 20082 V Minetest: 2018-03-15 20:49:44: VERBOSE[ConnectionSend]: con(43/2440)RunTimeouts(): Removing peer 1
03-15 20:49:44.561 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: Audio: Deinitializing...
03-15 20:49:44.591 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: Audio: Deinitialized.
03-15 20:49:44.601 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: ~TextureSource() 25/20
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: Irrlicht resources after cleanup:
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 	Remaining meshes   : 0
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 	Remaining textures : 20
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		0:#DefaultFont
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		1:/storage/emulated/0/Minetest/textures/base/pack/blank.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		2:/storage/emulated/0/Minetest/textures/base/pack/blank.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		3:/storage/emulated/0/Minetest/textures/base/pack/server_flags_creative.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		4:/storage/emulated/0/Minetest/textures/base/pack/server_flags_creative.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		5:/storage/emulated/0/Minetest/textures/base/pack/server_flags_damage.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		6:/storage/emulated/0/Minetest/textures/base/pack/server_flags_damage.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		7:/storage/emulated/0/Minetest/textures/base/pack/server_flags_favorite.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		8:/storage/emulated/0/Minetest/textures/base/pack/server_flags_favorite.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		9:/storage/emulated/0/Minetest/textures/base/pack/server_flags_pvp.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		10:/storage/emulated/0/Minetest/textures/base/pack/server_flags_pvp.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		11:/storage/emulated/0/Minetest/textures/base/pack/server_ping_1.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		12:/storage/emulated/0/Minetest/textures/base/pack/server_ping_1.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		13:/storage/emulated/0/Minetest/textures/base/pack/server_ping_2.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		14:/storage/emulated/0/Minetest/textures/base/pack/server_ping_2.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		15:/storage/emulated/0/Minetest/textures/base/pack/server_ping_3.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		16:/storage/emulated/0/Minetest/textures/base/pack/server_ping_3.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		17:/storage/emulated/0/Minetest/textures/base/pack/server_ping_4.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		18:/storage/emulated/0/Minetest/textures/base/pack/server_ping_4.png
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 		19:TTFontGlyphPage_Arimo.Regular.35_0
03-15 20:49:44.611 19970 19983 D Minetest: 2018-03-15 20:49:44: INFO[Main]: 	Remaining materials: 24 (note: irrlicht doesn't support removing renderers)
03-15 20:49:44.611 19970 19983 F libc    : Fatal signal 11 (SIGSEGV), code 1, fault addr 0xd2 in tid 19983 (MinetestNativeT)
03-15 20:49:44.701  2185  2185 F DEBUG   : pid: 19970, tid: 19983, name: MinetestNativeT  >>> net.minetest.minetest <<<
03-15 20:49:45.581  3893  3893 E audit   : type=1701 msg=audit(1521121785.581:12860): auid=4294967295 uid=10179 gid=10179 ses=4294967295 subj=u:r:untrusted_app:s0:c512,c768 pid=19983 comm="MinetestNativeT" reason="memory violation" sig=11
```

**Comment by srifqi** [comment]
### Attempt 2
I've fixed the SIGSEGV error. Now, it **doesn't respond** to touch screen input.

This is the `diff` to fix SIGSEGV error:

```diff
diff --git a/src/gui/touchscreengui.cpp b/src/gui/touchscreengui.cpp
index 9b573165..043695c2 100644
--- a/src/gui/touchscreengui.cpp
+++ b/src/gui/touchscreengui.cpp
@@ -376,8 +376,10 @@ void AutoHideButtonBar::deactivate()
 void AutoHideButtonBar::hide()
 {
 	m_visible = false;
-	m_starter.guibutton->setVisible(false);
-	m_starter.guibutton->setEnabled(false);
+	if (m_starter.guibutton) {
+		m_starter.guibutton->setVisible(false);
+		m_starter.guibutton->setEnabled(false);
+	}
 
 	std::vector<button_info*>::iterator iter = m_buttons.begin();
 
@@ -1103,17 +1105,17 @@ TouchScreenGUI::~TouchScreenGUI()
 		}
 	}
 
-	if (m_joystick_btn_off->guibutton) {
+	if (m_joystick_btn_off && m_joystick_btn_off->guibutton) {
 		m_joystick_btn_off->guibutton->drop();
 		m_joystick_btn_off->guibutton = NULL;
 	}
 
-	if (m_joystick_btn_bg->guibutton) {
+	if (m_joystick_btn_bg && m_joystick_btn_bg->guibutton) {
 		m_joystick_btn_bg->guibutton->drop();
 		m_joystick_btn_bg->guibutton = NULL;
 	}
 
-	if (m_joystick_btn_center->guibutton) {
+	if (m_joystick_btn_center && m_joystick_btn_center->guibutton) {
 		m_joystick_btn_center->guibutton->drop();
 		m_joystick_btn_center->guibutton = NULL;
 	}
@@ -1203,7 +1205,7 @@ void TouchScreenGUI::Toggle(bool visible)
 		}
 	}
 
-	if (m_joystick_btn_off->guibutton) {
+	if (m_joystick_btn_off && m_joystick_btn_off->guibutton) {
 		m_joystick_btn_off->guibutton->setVisible(visible);
 	}
 
```

**Comment by stujones11** [comment]
I think the problem here is that neither `guiConfirmRegistration` or `guiPasswordChange` have been written with any consideration for touch-screen devices, see `guiFormSpecMenu` as a comparison.

I also find it strange that commenting out lines in `touchsceengui` could fix the SIGSEGV error as the global pointer `g_touchscreengui` is not actually initialized until the game begins, perhaps that's the reason but I can't seem to find where these methods are being called for. The fact that the destructor is also being called would suggest that something went very wrong.

**Comment by srifqi** [comment]
@stujones11 I realized that but I had no idea how to use lua formspec in that code. Do you want to try? :)

**Comment by stujones11** [comment]
> Do you want to try?

I would have no idea how you could do this with formspecs or even if it is possible. It would certainly save a lot of trouble if it were.

I have found that you can prevent the crash by moving the touchscreen initialization from `Game::initGui` to `Game::createClient` before the server the server connection is made.
https://github.com/minetest/minetest/blob/8d3190e77a9d13bf0bca3ecdf40a788ce1d39d58/src/game.cpp#L1413-L1418
It's a bit hacky but does confirm my suspicion that it is be accessed form somewhere before it has been initialized. It would be better to find out where that is, though it might not even be confirm dialog to blame for this.

The next problem we have is making the dialog respond to touch input. The code for this can be found in `guiFormSpecMenu` but simply copying that would result in a great deal of code duplication, considering this also needs done for the password change menu as well as any future dialogs.

The most logical solution (IMO) would be to move the touchscreen event translation and native input methods into the base class `GUIModalMenu`. It will be quite a bit of work but maybe worth it in the long-run. I might have a go over the weekend if no one beats me to it ;)

**Comment by Ezhh** [comment]
Can this be disabled for Android so it doesn't need to be a blocker? (Label or not, it is really, really bad to release like this).

(Personally I'd prefer the whole thing be an option you enable if you are someone who tends to mess up your passwords, but I'm guessing this is too much to ask for.)

**Comment by paramat** [comment]
Ezhh the fixing PR for this #7483 also fixes another blocker issue. So although i would agree to removing the server account confirmation it looks like this will get fixed anyway.

**Comment by paramat** [comment]
Fixed by #7483
